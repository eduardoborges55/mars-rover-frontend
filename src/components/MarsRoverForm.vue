<template>
  <div>
    <h2>Configurações do Planalto e Rover</h2>

    <label for="plateauSize">Tamanho do Planalto:</label>
    <input type="text" id="plateauSize" v-model="plateauSize" placeholder="Ex: 5 5" />

    <label for="roverPosition">Posição do Rover:</label>
    <input type="text" id="roverPosition" v-model="roverPosition" placeholder="Ex: 1 2 N" />

    <label for="roverInstructions">Instruções do Rover:</label>
    <input type="text" id="roverInstructions" v-model="roverInstructions" placeholder="Ex: LMLMLMLMM" />

    <button @click="submitForm">Enviar</button>

    <p v-if="error">{{ error }}</p>
  </div>
</template>

<script>
import axios from '@/axios';

export default {
  data() {
    return {
      plateauSize: '',
      roverPosition: '',
      roverInstructions: '',
      previousFormData: null,
      error: '', 
    };
  },
  methods: {
    async submitForm() {
      try {
        this.error = '';

        if (!this.plateauSize || !this.roverPosition || !this.roverInstructions) {
          throw new Error('Todos os campos são obrigatórios.');
        }
        const roverPositionArray = this.roverPosition.split(' ');

        const roverData = {
          x: parseInt(roverPositionArray[0]),
          y: parseInt(roverPositionArray[1]),
          direction: roverPositionArray[2],
          instructions: this.roverInstructions,
        };

        if (JSON.stringify(roverData) === JSON.stringify(this.previousFormData)) {
          throw new Error('Os dados não foram alterados.');
        }

        this.previousFormData = { ...roverData };

        const response = await axios.post('/api/explore', {
          upperRightCoordinates: this.plateauSize.split(' ').map(Number),
          rovers: [roverData],
        });

        this.$emit('formSubmitted', response);
      } catch (error) {
        console.error('Erro ao enviar dados:', error.message);
        this.error = error.message;
      }
    },
  },
};
</script>

<style scoped>
</style>
