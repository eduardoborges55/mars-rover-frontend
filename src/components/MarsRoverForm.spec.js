import { mount } from '@vue/test-utils';
import MarsRoverForm from '@/components/MarsRoverForm.vue';
import mock from '../../jest.setup'
/* eslint-disable */
describe('MarsRoverForm.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(MarsRoverForm);
  });

  afterEach(() => {
    mock.reset();
  });

  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('submits the form', async () => {
   
    mock.onPost('/api/explore').reply(200, { data: {} });
  
    await wrapper.setData({
      plateauSize: '5 5',
      roverPosition: '1 2 N',
      roverInstructions: 'LMLMLMLMM',
    });
  
    const submitFormSpy = jest.spyOn(wrapper.vm, 'submitForm');
 
    await wrapper.vm.submitForm();
  
    expect(submitFormSpy).toHaveBeenCalled();
  
    const postData = JSON.parse(mock.history.post[0].data);
    expect(mock.history.post.length).toBe(1);
    expect(mock.history.post[0].url).toBe('/api/explore');
    expect(postData).toEqual({
      upperRightCoordinates: [5, 5],
      rovers: [{
        x: 1,
        y: 2,
        direction: 'N',
        instructions: 'LMLMLMLMM',
      }],
    });
  });
});
/* eslint-enable */