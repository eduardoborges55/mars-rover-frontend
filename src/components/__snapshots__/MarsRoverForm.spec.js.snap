// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarsRoverForm.vue renders correctly 1`] = `
<div>
  <h2>
    Configurações do Planalto e Rover
  </h2>
   
  <label
    for="plateauSize"
  >
    Tamanho do Planalto:
  </label>
   
  <input
    id="plateauSize"
    placeholder="Ex: 5 5"
    type="text"
  />
   
  <label
    for="roverPosition"
  >
    Posição do Rover:
  </label>
   
  <input
    id="roverPosition"
    placeholder="Ex: 1 2 N"
    type="text"
  />
   
  <label
    for="roverInstructions"
  >
    Instruções do Rover:
  </label>
   
  <input
    id="roverInstructions"
    placeholder="Ex: LMLMLMLMM"
    type="text"
  />
   
  <button>
    Enviar
  </button>
   
  <!---->
</div>
`;
